Move go-elektra into bindings folder

tmakar opened the pull request:
<!--
Check relevant points but **please do not remove entries**.
-->

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add them to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - add a line in `doc/news/_preparation_next_release.md`
  - reformat the code with `scripts/dev/reformat-all`
  - make all unit tests pass
  - fix all memleaks
  - fix the CI itself (or rebase if already fixed)
- [ ] The PR is rebased with current master.

<!--
If you have any troubles fulfilling these criteria, please write about the trouble as comment in the PR.
We will help you, but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For documentation fixes, spell checking, and similar none of these points below need to be checked.
-->

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

<!--
Reviewers should check the following.
-->

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

<!--
If you are already Elektra developer, please adjust the labels.
Otherwise, write a comment and it will be done for you.
-->

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if everything is done** and no further pushes are planned by you.

atmaxinger commented:
I'm not really qualified enough to review anything related to Go, but I think you also need to modify some CMakeLists.txt files so that it actually builds with the rest of Elektra.

atmaxinger commented:
For starters, you may find the CMakeLists.txt files for the [Rust](https://github.com/ElektraInitiative/libelektra/blob/master/src/bindings/rust/CMakeLists.txt) and [Java](https://github.com/ElektraInitiative/libelektra/blob/master/src/bindings/jna/CMakeLists.txt) bindings helpful.

atmaxinger commented:
Maybe also rename the folder from `go-elektra` to just `go` to be consistent with the other binding.

atmaxinger commented on src/bindings/go-elektra/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+
+	add_custom_command (COMMAND ${GO_EXECUTABLE} build ./kdb WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
+else ()
+	exclude_binding (go-elektra, "elektra not found")
```
I think you meant "go not found" 😉

tmakar replied:
Yes, you are right 👍

hannes99 reviewed (approved):
Looks good to me 👍

atmaxinger commented on src/bindings/go-elektra/.travis.yml:
```diff
@@ -0,0 +1,13 @@
+dist: focal
```
I think this file is pretty useless, as `.travis.yml` would need to be at the root directory of the repository.

atmaxinger reviewed (changes requested):
Great work, thanks! However, you must still add the go compiler to all of our CI Docker images (`scripts/docker`) and other CI definitions (`.github/workflows`, `.cirrus.yml`) . As of now, the binding isn't built in our CI.

You need to add them to the Docker images at least. Also GitHub workflows would be logical, as this is our macOS build.

Maybe @mpranj or @markus2330 could quickly comment on where you have to add it.

markus2330 commented:
Yes, at least in one of our Docker images (scripts/docker), go should be installed so that the binding and elektrad can be built and tested.

atmaxinger commented:
It seems like the OpenSUSE docker image couldn't be built.

tmakar commented:
> It seems like the OpenSUSE docker image couldn't be built.

The `\` after the package name was causing the issue.

atmaxinger commented:
jenkins build libelektra please

atmaxinger commented:
Seems like there is a permission error on the Fedora build related to go

```
failed to initialize build cache at /home/jenkins/.cache/go-build: mkdir /home/jenkins/.cache/go-build: permission denied
make[2]: *** [src/tools/elektrad/CMakeFiles/elektrad.dir/build.make:74: src/tools/elektrad/elektrad/elektrad] Error 1
make[1]: *** [CMakeFiles/Makefile2:14219: src/tools/elektrad/CMakeFiles/elektrad.dir/all] Error 2
```

tmakar commented:
jenkins build libelektra please

tmakar commented:
jenkins build libelektra please

tmakar commented:
jenkins build libelektra please

tmakar commented:
jenkins build libelektra please

tmakar commented:
jenkins build libelektra please

tmakar commented:
jenkins build libelektra please

tmakar commented:
jenkins build libelektra please

atmaxinger commented:
jenkins build libelektra please

atmaxinger commented:
Build error on Debian:

```
go build go.libelektra.org/kdb: invalid flag in pkg-config --cflags: directory/src/include
```

tmakar commented:
@atmaxinger Is there some registry I need to push the docker images?

tmakar commented:
Ok, I see, it is done in the Jenkins pipeline.

tmakar commented:
jenkins build libelektra please

tmakar commented:
jenkins build libelektra please

tmakar commented:
```text
! Missing $ inserted.
<inserted text> 
                $
l.872 \end{DoxyRetVals}
? 
! Emergency stop.
<inserted text> 
                $
l.872 \end{DoxyRetVals}
!  ==> Fatal error occurred, no output PDF file produced!
```

@atmaxinger @flo91 

What I can extract from the `doc` folders `CMakeLists.txt` file, this process of building a pdf is done there. Is this error may due to some wrong documentation in the `_preparation_next_release.md`?

atmaxinger commented:
No this is because of some weird error when converting the Doxygen API docs to PDF. We had this problem in the past too on Debian `sid`. [Here is the bug report](https://github.com/doxygen/doxygen/issues/9567). AFAIR we just disabled building in Debian `sid` because we couldn't find any other way of dealing with it. But you might try the solutions proposed in the issue too.

tmakar commented:
> No this is because of some weird error when converting the Doxygen API docs to PDF. We had this problem in the past too on Debian `sid`. [Here is the bug report](https://github.com/doxygen/doxygen/issues/9567). AFAIR we just disabled building in Debian `sid` because we couldn't find any other way of dealing with it. But you might try the solutions proposed in the issue too.

Thank you!

tmakar commented:
@atmaxinger Do you maybe know what this is about?

```sh
Sending interrupt signal to process
Terminated
make[2]: *** Deleting file 'src/bindings/cpp/tests/CMakeFiles/testcpp_contextual_thread.dir/testcpp_contextual_thread.cpp.o'
make[2]: *** [src/libs/tools/src/CMakeFiles/elektratools.dir/build.make:438: src/libs/tools/src/CMakeFiles/elektratools.dir/plugins.cpp.o] Terminated
make[2]: *** [src/bindings/cpp/tests/CMakeFiles/testcpp_contextual_thread.dir/build.make:78: src/bindings/cpp/tests/CMakeFiles/testcpp_contextual_thread.dir/testcpp_contextual_thread.cpp.o] Terminated
make[2]: *** wait: No child processes.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** wait: No child processes.  Stop.
make[2]: *** Deleting file 'src/bindings/cpp/tests/CMakeFiles/testcpp_ks.dir/testcpp_ks.cpp.o'
make[2]: *** [src/bindings/cpp/tests/CMakeFiles/testcpp_key.dir/build.make:78: src/bindings/cpp/tests/CMakeFiles/testcpp_key.dir/testcpp_key.cpp.o] Terminated
make[2]: *** [src/libs/tools/src/CMakeFiles/elektratools-full.dir/build.make:414: src/libs/tools/src/CMakeFiles/elektratools-full.dir/plugin.cpp.o] Terminated
make[2]: *** [src/libs/tools/src/CMakeFiles/elektratools-full.dir/build.make:426: src/libs/tools/src/CMakeFiles/elektratools-full.dir/plugindatabase.cpp.o] Terminated
make[2]: *** [src/libs/tools/src/CMakeFiles/elektratools-full.dir/build.make:438: src/libs/tools/src/CMakeFiles/elektratools-full.dir/plugins.cpp.o] Terminated
make[2]: *** [src/libs/tools/src/CMakeFiles/elektratools-static.dir/build.make:438: src/libs/tools/src/CMakeFiles/elektratools-static.dir/plugins.cpp.o] Terminated
make[1]: *** [CMakeFiles/Makefile2:11415: src/libs/tools/src/CMakeFiles/elektratools.dir/all] Terminated
make[2]: *** [src/bindings/cpp/benchmarks/CMakeFiles/benchmark_sync.dir/build.make:78: src/bindings/cpp/benchmarks/CMakeFiles/benchmark_sync.dir/benchmark_sync.cpp.o] Terminated
make[1]: *** [CMakeFiles/Makefile2:13284: src/bindings/cpp/tests/CMakeFiles/testcpp_contextual_basic.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:13508: src/bindings/cpp/tests/CMakeFiles/testcpp_contextual_thread.dir/all] Terminated
make[1]: *** [CMakeFiles/Makefile2:13668: src/bindings/cpp/tests/CMakeFiles/testcpp_key.dir/all] Terminated
make[2]: *** [src/libs/tools/src/CMakeFiles/elektratools-static.dir/build.make:462: src/libs/tools/src/CMakeFiles/elektratools-static.dir/specreader.cpp.o] Terminated
make[2]: *** [src/bindings/cpp/tests/CMakeFiles/testcpp_ks.dir/build.make:78: src/bindings/cpp/tests/CMakeFiles/testcpp_ks.dir/testcpp_ks.cpp.o] Terminated
make[2]: *** [src/libs/tools/src/CMakeFiles/elektratools-full.dir/build.make:462: src/libs/tools/src/CMakeFiles/elektratools-full.dir/specreader.cpp.o] Terminated
make[2]: *** [src/bindings/cpp/tests/CMakeFiles/testcpp_contextual_update.dir/build.make:78: src/bindings/cpp/tests/CMakeFiles/testcpp_contextual_update.dir/testcpp_contextual_update.cpp.o] Terminated
make[2]: *** [src/bindings/cpp/benchmarks/CMakeFiles/benchmark_thread.dir/build.make:78: src/bindings/cpp/benchmarks/CMakeFiles/benchmark_thread.dir/benchmark_thread.cpp.o] Terminated
make[1]: *** [CMakeFiles/Makefile2:11357: src/libs/tools/src/CMakeFiles/elektratools-static.dir/all] Terminated
make[1]: *** [CMakeFiles/Makefile2:11384: src/libs/tools/src/CMakeFiles/elektratools-full.dir/all] Terminated
make[1]: *** [CMakeFiles/Makefile2:13636: src/bindings/cpp/tests/CMakeFiles/testcpp_ks.dir/all] Terminated
make[1]: *** [CMakeFiles/Makefile2:14062: src/bindings/cpp/benchmarks/CMakeFiles/benchmark_thread.dir/all] Terminated
make[1]: *** [CMakeFiles/Makefile2:13540: src/bindings/cpp/tests/CMakeFiles/testcpp_contextual_update.dir/all] Terminated
make[1]: *** [CMakeFiles/Makefile2:14092: src/bindings/cpp/benchmarks/CMakeFiles/benchmark_sync.dir/all] Terminated
make: *** [Makefile:182: all] Terminated
script returned exit code 143
```

atmaxinger commented:
Please rebase this

atmaxinger commented on scripts/docker/centos/stream8/Dockerfile:
```diff
@@ -78,6 +79,13 @@ RUN dnf upgrade --refresh -y && dnf install -y \
+
+# see https://github.com/golang/go/wiki/InvalidFlag
+ENV CGO_CFLAGS_ALLOW="directory.*"
+ENV LD_LIBRARY_PATH=directory/lib:${LD_LIBRARY_PATH}
```
Are you sure that `directory` shouldn't be, like, an actual directory?

tmakar replied:
I planned to remove it, thanks for pointing it out! :+1:

tmakar commented:
@atmaxinger 

```sh
go build go.libelektra.org/kdb: invalid flag in pkg-config --cflags: directory/src/include
[ 72%] Built target testmod_spec
make[2]: *** [src/tools/elektrad/CMakeFiles/elektrad.dir/build.make:74: src/tools/elektrad/elektrad/elektrad] Error 1
make[1]: *** [CMakeFiles/Makefile2:14219: src/tools/elektrad/CMakeFiles/elektrad.dir/all] Error 2
```
The directory is really called `directory`.

tmakar commented:
@markus2330 @atmaxinger I will put this on hold as I'm running out of ideas here :/

markus2330 commented:
Please write (maybe in an issue or here) what exactly the problem seems to be. The current problem from the build server seems to be quite trivial to solve:

`go build go.libelektra.org/kdb: invalid flag in pkg-config --libs: directory/lib`

Essential for fixing something like this is that you have a proper setup locally where you can debug it.

tmakar commented:
I think I have found a possible workaround for this.

> Please write (maybe in an issue or here) what exactly the problem seems to be. The current problem from the build server seems to be quite trivial to solve:
> 
> `go build go.libelektra.org/kdb: invalid flag in pkg-config --libs: directory/lib`
> 
> Essential for fixing something like this is that you have a proper setup locally where you can debug it.

I think I have found a possible workaround for this.

tmakar commented:
This is what I get in the CirrusCI (@markus2330, @atmaxinger) but I didn't change anything in c code:

```sh
FAILED: src/libs/tools/src/CMakeFiles/elektratools.dir/helper/keyhelper.cpp.o 
/usr/sbin/c++ -DELEKTRA_SHARED -DHAVE_KDBCONFIG_H -Delektratools_EXPORTS -I/tmp/cirrus-ci-build/build/src/include -I/tmp/cirrus-ci-build/src/include -I/tmp/cirrus-ci-build/build/src/bindings/cpp/include -I/tmp/cirrus-ci-build/src/bindings/cpp/include -I/tmp/cirrus-ci-build/build/src/libs/tools/include -I/tmp/cirrus-ci-build/src/libs/tools/include -std=c++11  -Wno-deprecated-declarations  -Wold-style-cast -Wstrict-null-sentinel -D_GLIBCXX_USE_NANOSLEEP -Wno-missing-field-initializers -Woverloaded-virtual  -Wsign-promo  -Wno-long-long -Wpedantic -Wno-variadic-macros -Wall -Wextra -Wno-overlength-strings -Wsign-compare -Wfloat-equal -Wformat -Wformat-security -Wshadow -Wcomments -Wtrigraphs -Wundef -Wuninitialized -Winit-self -Wmaybe-uninitialized -O2 -g -DNDEBUG -fPIC -MD -MT src/libs/tools/src/CMakeFiles/elektratools.dir/helper/keyhelper.cpp.o -MF src/libs/tools/src/CMakeFiles/elektratools.dir/helper/keyhelper.cpp.o.d -o src/libs/tools/src/CMakeFiles/elektratools.dir/helper/keyhelper.cpp.o -c /tmp/cirrus-ci-build/src/libs/tools/src/helper/keyhelper.cpp
In file included from /tmp/cirrus-ci-build/src/bindings/cpp/include/kdbio.hpp:18,
                 from /tmp/cirrus-ci-build/src/bindings/cpp/include/kdbexcept.hpp:23,
                 from /tmp/cirrus-ci-build/src/bindings/cpp/include/kdb.hpp:15,
                 from /tmp/cirrus-ci-build/src/libs/tools/include/helper/keyhelper.hpp:13,
                 from /tmp/cirrus-ci-build/src/libs/tools/src/helper/keyhelper.cpp:10:
/tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:42:6: warning: elaborated-type-specifier for a scoped enum must not use the 'class' keyword
   42 | enum class ElektraNamespace : std::uint8_t
      | ~~~~ ^~~~~
      |      -----
/tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:42:12: error: use of enum 'ElektraNamespace' without previous declaration
   42 | enum class ElektraNamespace : std::uint8_t
      |            ^~~~~~~~~~~~~~~~
/tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:42:29: error: expected unqualified-id before ':' token
   42 | enum class ElektraNamespace : std::uint8_t
      |                             ^
/tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:55:14: error: 'ElektraNamespace' does not name a type; did you mean 'elektraNamespace'?
   55 | static const ElektraNamespace ELEKTRA_NAMESPACES[] = {
      |              ^~~~~~~~~~~~~~~~
      |              elektraNamespace
/tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:235:16: error: 'ElektraNamespace' does not name a type; did you mean 'elektraNamespace'?
  235 |         inline ElektraNamespace getNamespace () const;
      |                ^~~~~~~~~~~~~~~~
      |                elektraNamespace
/tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:236:38: error: 'ElektraNamespace' has not been declared
  236 |         inline ssize_t setNamespace (ElektraNamespace ns) const;
```

tmakar commented:
```sh
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/google/uuid v1.3.0
go: downloading go.libelektra.org v0.0.0-20230321081411-7a7dc1493aee
# go.libelektra.org/kdb
gcc: error: directory/src/include: No such file or directory
make[2]: *** [src/tools/elektrad/CMakeFiles/elektrad.dir/build.make:81: src/tools/elektrad/elektrad/elektrad] Error 2
make[1]: *** [CMakeFiles/Makefile2:15409: src/tools/elektrad/CMakeFiles/elektrad.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

atmaxinger commented:
> gcc: error: directory/src/include: No such file or directory

I'm going to say it again: Somewhere in your configuration you have the string `directory`, where a real path should be.

tmakar commented:
When running this locally, I don't get any errors :expressionless:

tmakar commented:
> > gcc: error: directory/src/include: No such file or directory
> 
> I'm going to say it again: Somewhere in your configuration you have the string `directory`, where a real path should be.

Would be nice if this is the case 😅 unfortunately, I couldn‘t find it anywhere which is very frustrating, as it seems that this is detected at runtime, which also does not make much sense ( I mean, why would it use „directory“ at runtime ?!?!) 😵‍💫

tmakar commented:
Maybe it’s connected to the cache workaround in docker for golang, which I have used. I will check that.

tmakar commented:
jenkins build libelektra please

tmakar commented:
jenkins build libelektra please

tmakar commented:
@markus2330 @atmaxinger 

I have now checked and it seems that it fails just for debian bulleye and fedora 37. For all the CirrusCI phases it works perfect (example is Fedora `/scripts/docker/cirrus/fedora/Dockerfile`).

I will investigate further why this is not working for bulleye and fedora on jenkins.

tmakar commented:
@markus2330 @atmaxinger 

I have now found that it only occurs on `jenkins` in general, on `CirrusCI` it works perfectly.

Seems that `${buildDir}` is getting set to `directory` when run on jenkins.

tmakar commented:
@markus2330 Do you think we can agree that I remove `elektrad` pipeline builds etc. for all jenkins phases and leave it for the `CirrusCI` ?

tmakar commented:
We are currently experiencing some issues with jenkins as well:

```sh
Starting pull/hub.libelektra.org/build-elektra-debian-buster-minimal:202305-94de7aed5e0ec5c307ced7f13a6bec38797bee7bfcbe94cf25faf2b16e2fbf6c/ on a7
/home/jenkins/workspace/libelektra_PR-4895@tmp/ec44003d-f23a-46f6-bfcf-61ae3bf17651: No space left on device
```

tmakar commented:
@atmaxinger go is now moved into the repo.

Build & tests for `elektrad` tool and `go-elektra` binding, run on Cirrus CI (successfully).

tmakar commented:
jenkins build libelektra please

tmakar commented:
@atmaxinger @kodebach 

Have you maybe  encountered such an issue before?


```
FAILED: src/libs/tools/src/CMakeFiles/elektratools.dir/helper/keyhelper.cpp.o 
/usr/sbin/c++ -DELEKTRA_SHARED -DHAVE_KDBCONFIG_H -Delektratools_EXPORTS -I/tmp/cirrus-ci-build/build/src/include -I/tmp/cirrus-ci-build/src/include -I/tmp/cirrus-ci-build/build/src/bindings/cpp/include -I/tmp/cirrus-ci-build/src/bindings/cpp/include -I/tmp/cirrus-ci-build/build/src/libs/tools/include -I/tmp/cirrus-ci-build/src/libs/tools/include -std=c++11  -Wno-deprecated-declarations  -Wold-style-cast -Wstrict-null-sentinel -D_GLIBCXX_USE_NANOSLEEP -Wno-missing-field-initializers -Woverloaded-virtual  -Wsign-promo  -Wno-long-long -Wpedantic -Wno-variadic-macros -Wall -Wextra -Wno-overlength-strings -Wsign-compare -Wfloat-equal -Wformat -Wformat-security -Wshadow -Wcomments -Wtrigraphs -Wundef -Wuninitialized -Winit-self -Wmaybe-uninitialized -O2 -g -DNDEBUG -fPIC -MD -MT src/libs/tools/src/CMakeFiles/elektratools.dir/helper/keyhelper.cpp.o -MF src/libs/tools/src/CMakeFiles/elektratools.dir/helper/keyhelper.cpp.o.d -o src/libs/tools/src/CMakeFiles/elektratools.dir/helper/keyhelper.cpp.o -c /tmp/cirrus-ci-build/src/libs/tools/src/helper/keyhelper.cpp
In file included from /tmp/cirrus-ci-build/src/bindings/cpp/include/kdbio.hpp:18,
                 from /tmp/cirrus-ci-build/src/bindings/cpp/include/kdbexcept.hpp:23,
                 from /tmp/cirrus-ci-build/src/bindings/cpp/include/kdb.hpp:15,
                 from /tmp/cirrus-ci-build/src/libs/tools/include/helper/keyhelper.hpp:13,
                 from /tmp/cirrus-ci-build/src/libs/tools/src/helper/keyhelper.cpp:10:
/tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:42:6: warning: elaborated-type-specifier for a scoped enum must not use the 'class' keyword
   42 | enum class ElektraNamespace : std::uint8_t
      | ~~~~ ^~~~~
      |      -----
/tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:42:12: error: use of enum 'ElektraNamespace' without previous declaration
   42 | enum class ElektraNamespace : std::uint8_t
      |            ^~~~~~~~~~~~~~~~
/tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:42:29: error: expected unqualified-id before ':' token
   42 | enum class ElektraNamespace : std::uint8_t
      |                             ^
/tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:55:14: error: 'ElektraNamespace' does not name a type; did you mean 'elektraNamespace'?
   55 | static const ElektraNamespace ELEKTRA_NAMESPACES[] = {
      |              ^~~~~~~~~~~~~~~~
      |              elektraNamespace
/tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:235:16: error: 'ElektraNamespace' does not name a type; did you mean 'elektraNamespace'?
  235 |         inline ElektraNamespace getNamespace () const;
      |                ^~~~~~~~~~~~~~~~
      |                elektraNamespace
/tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:236:38: error: 'ElektraNamespace' has not been declared
  236 |         inline ssize_t setNamespace (ElektraNamespace ns) const;
```

tmakar commented:
@markus2330 @atmaxinger

I think I have found the problem with jenkins :exploding_head: 

As it was working on `CirrusCI` and I also tested it locally by pulling the exact same docker image which was used on jenkins and it worked locally, I investigated the `Jenkinsfile`.

Yesterday, I found this piece of code in `scripts/jenkins/Jenkinsfile`:
```groovy
return [(testName): {
    stage(testName) {
      withDockerEnv(image, dockerOpts) {
        // we use a space in the directory to test if paths are
        // properly escaped
        def buildDir='build directory'

        if(tests) {
          artifacts.add("\"${buildDir}\"/Testing/*/*.xml")
        }

        try {
```

Now the interesting part:
`Go` seems to have issue with spaces in any folders etc. so it fails on builds / tests, etc. As we are using `build directory` as our build directory and it has a space in the name it will lead to issues with `go`. I have removed this whitespace now.

Currently the build fails because we ran out of disk (don't know why but it seemed to work for a short time today, maybe someone fixed it silently :smile:).

tmakar commented:
> @atmaxinger @kodebach
> 
> Have you maybe encountered such an issue before?
> 
> ```
> FAILED: src/libs/tools/src/CMakeFiles/elektratools.dir/helper/keyhelper.cpp.o 
> /usr/sbin/c++ -DELEKTRA_SHARED -DHAVE_KDBCONFIG_H -Delektratools_EXPORTS -I/tmp/cirrus-ci-build/build/src/include -I/tmp/cirrus-ci-build/src/include -I/tmp/cirrus-ci-build/build/src/bindings/cpp/include -I/tmp/cirrus-ci-build/src/bindings/cpp/include -I/tmp/cirrus-ci-build/build/src/libs/tools/include -I/tmp/cirrus-ci-build/src/libs/tools/include -std=c++11  -Wno-deprecated-declarations  -Wold-style-cast -Wstrict-null-sentinel -D_GLIBCXX_USE_NANOSLEEP -Wno-missing-field-initializers -Woverloaded-virtual  -Wsign-promo  -Wno-long-long -Wpedantic -Wno-variadic-macros -Wall -Wextra -Wno-overlength-strings -Wsign-compare -Wfloat-equal -Wformat -Wformat-security -Wshadow -Wcomments -Wtrigraphs -Wundef -Wuninitialized -Winit-self -Wmaybe-uninitialized -O2 -g -DNDEBUG -fPIC -MD -MT src/libs/tools/src/CMakeFiles/elektratools.dir/helper/keyhelper.cpp.o -MF src/libs/tools/src/CMakeFiles/elektratools.dir/helper/keyhelper.cpp.o.d -o src/libs/tools/src/CMakeFiles/elektratools.dir/helper/keyhelper.cpp.o -c /tmp/cirrus-ci-build/src/libs/tools/src/helper/keyhelper.cpp
> In file included from /tmp/cirrus-ci-build/src/bindings/cpp/include/kdbio.hpp:18,
>                  from /tmp/cirrus-ci-build/src/bindings/cpp/include/kdbexcept.hpp:23,
>                  from /tmp/cirrus-ci-build/src/bindings/cpp/include/kdb.hpp:15,
>                  from /tmp/cirrus-ci-build/src/libs/tools/include/helper/keyhelper.hpp:13,
>                  from /tmp/cirrus-ci-build/src/libs/tools/src/helper/keyhelper.cpp:10:
> /tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:42:6: warning: elaborated-type-specifier for a scoped enum must not use the 'class' keyword
>    42 | enum class ElektraNamespace : std::uint8_t
>       | ~~~~ ^~~~~
>       |      -----
> /tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:42:12: error: use of enum 'ElektraNamespace' without previous declaration
>    42 | enum class ElektraNamespace : std::uint8_t
>       |            ^~~~~~~~~~~~~~~~
> /tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:42:29: error: expected unqualified-id before ':' token
>    42 | enum class ElektraNamespace : std::uint8_t
>       |                             ^
> /tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:55:14: error: 'ElektraNamespace' does not name a type; did you mean 'elektraNamespace'?
>    55 | static const ElektraNamespace ELEKTRA_NAMESPACES[] = {
>       |              ^~~~~~~~~~~~~~~~
>       |              elektraNamespace
> /tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:235:16: error: 'ElektraNamespace' does not name a type; did you mean 'elektraNamespace'?
>   235 |         inline ElektraNamespace getNamespace () const;
>       |                ^~~~~~~~~~~~~~~~
>       |                elektraNamespace
> /tmp/cirrus-ci-build/src/bindings/cpp/include/key.hpp:236:38: error: 'ElektraNamespace' has not been declared
>   236 |         inline ssize_t setNamespace (ElektraNamespace ns) const;
> ```

found it already and fixed, no need to respond :smiley:

atmaxinger commented:
> found it already and fixed, no need to respond 😃

What was it?

tmakar commented:
> > found it already and fixed, no need to respond 😃
> 
> What was it?

In the dockerfiles which are used by cirrus I had env variables which are missed to remove it (I added them while doing some golang „testing“). however, I am not 100% sure why this affected explicitly the ElektraNamespaces.

kodebach commented:
I'm guessing it had something to do with `CGO_CFLAGS_ALLOW` and `CGO_LDFLAGS_ALLOW`.

> `Go` seems to have issue with spaces in any folder names etc. so it fails on builds / tests, etc.

Is it Go itself the problem or just our setup for the binding? The idea behind `'build directory'` was exactly that it would detect mistakes in our scripts (e.g. `$SOME_PATH` instead of `"$SOME_PATH"` in a shell script).

However, it would be very much in-line with what I've learned about the authors of Go, if their answer to "spaces in file paths" really was "just rename/move the file".

kodebach commented on src/bindings/go-elektra/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+	# build
+	add_custom_command (
+		OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/lib/libelektra.golib"
+		COMMAND ${CMAKE_COMMAND} ARGS -E env GO111MODULE=on PKG_CONFIG_PATH=${CMAKE_CURRENT_BINARY_DIR} ${GO_EXECUTABLE} build ./kdb
+		WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
```
```suggestion
		COMMAND ${CMAKE_COMMAND} ARGS -E env GO111MODULE=on PKG_CONFIG_PATH="${CMAKE_CURRENT_BINARY_DIR}" ${GO_EXECUTABLE} build ./kdb
		WORKING_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}")
```

Those are exactly the places where a space in the build directory path could cause a problem...

tmakar replied:
The error existed also before I added this…

kodebach replied:
Also `PKG_CONFIG_PATH=${CMAKE_CURRENT_BINARY_DIR}` won't help in any way AFAICT. There is no `elektra.pc` in this directory...

tmakar replied:
removed it!

kodebach replied:
> The error existed also before I added this…

It could also be that `elektra.pc` is simply generated incorrectly in `scripts/cmake/CMakeLists.txt`. Could you try building in a directory with a space and then check what the generated file `"build dir/scripts/cmake/elektra.pc"` looks like?

You probably also need to use something like `cmake -DCMAKE_INSTALL_PREFIX="$PWD/install" ..` to show the error.

tmakar replied:
Will try.

tmakar replied:
I have replaced my real `PWD` with `/my/path/to/libelektra/` here in this comment:

```text
prefix=/my/path/to/libelektra/build dir/install
exec_prefix=${prefix}/bin
libdir=${prefix}/lib
includedir=${prefix}/include/elektra
plugindir=${prefix}/lib/elektra
tool_execdir=${prefix}/lib/elektra/tool_exec
templatedir=${prefix}/share/elektra/templates
build_type=shared

Name: 
Description: 
Version: 0.9.14
Libs: -L${libdir} -lelektra-kdb -lelektra-core
Cflags: -I${includedir}
```

kodebach replied:
Yeah, that's what i suspected... Quotes won't work, but a `\` would:

```
prefix=/my/path/to/libelektra/build\ dir/install
```

So that means we need to fix the CMake script that generates this. Not sure how hard that is...

But I think this isn't actually what breaks in Jenkins. AFAICT Jenkins doesn't actually use a a subdirectory of the build dir as the `CMAKE_INSTALL_PREFIX`:

https://github.com/ElektraInitiative/libelektra/blob/d328680147620f5d78c00f547bf827287d2f6853/scripts/jenkins/Jenkinsfile#L60

So there is probably still some other place where the include dir is passed to the go stuff...

tmakar replied:
Just to note, it’s not just the include it also errors out for the linking library path.

i would suggest the following:
I will invest 1-2h more into finding a different solution, otherwise I would go with the solution I already committed yesterday (by not using a space in the dir name).

what do you say? Is this fine with you?

tmakar replied:
This is how the elektra.pc looks like which is used by elektrad which throws the error. Maybe fixing this `prefix=@CMAKE_BINARY_DIR@` helps? 

```
prefix=@CMAKE_BINARY_DIR@
libdir=${prefix}/lib
srcinclude=${prefix}/src/include
buildinclude=@CMAKE_SOURCE_DIR@/src/include

Name: @PACKAGE_NAME@
Description: @PACKAGE_DESCRIPTION@
Version: @KDB_VERSION@
Cflags: -I${srcinclude} -I${buildinclude}
Libs: -L${libdir} -lelektra-kdb -lelektr
```

tmakar replied:
I would leave it as it is now with the removed whitespace!

kodebach commented on src/bindings/go-elektra/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+	add_custom_command (
+		OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/tests/target/release"
+		COMMAND ${GO_EXECUTABLE} test ./kdb
+		WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
```
```suggestion
		WORKING_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}")
```

kodebach commented on src/bindings/go-elektra/LICENSE.md:
This file isn't really needed here anymore. This is the default license for the entire repo.

tmakar replied:
Yes, missed this, thanks!!

tmakar commented:
> I'm guessing it had something to do with `CGO_CFLAGS_ALLOW` and `CGO_LDFLAGS_ALLOW`.
> 
> > `Go` seems to have issue with spaces in any folder names etc. so it fails on builds / tests, etc.
> 
> Is it Go itself the problem or just our setup for the binding? The idea behind `'build directory'` was exactly that it would detect mistakes in our scripts (e.g. `$SOME_PATH` instead of `"$SOME_PATH"` in a shell script).
> 
> However, it would be very much in-line with what I've learned about the authors of Go, if their answer to "spaces in file paths" really was "just rename/move the file".

It’s really like that. I removed the `CGO_CFLAGS_ALLOW` and `CGO_LDFLAGS_ALLOW` envs also and tried it today but it didn‘t work. As I have seen the same error messages over and over again, I‘ve overseen that completely ignored the name before the space, so it tried to search for „directory/src/includes“ etc. which obviously does not exist, as the name is „build directory“ and not just „directory“. 😅

kodebach commented on src/bindings/go-elektra/kdb/kdb.go:
```diff
@@ -0,0 +1,160 @@
+package kdb
+
+// #cgo pkg-config: elektra
```
The immediate cause of the space issue is probably this. However, this line is not problematic and supports spaces properly (https://github.com/golang/go/issues/16455), as long as the pkg-config file is correct.

However, the bigger issue here is that this will not work, even if the `PKG_CONFIG_PATH` was set better in CMake. The issue is that `pkg-config` only finds an installed version of Elektra, it the pkgconfig files will never work properly for the version still being built. We'd have to generate special variants of the pkgconfig files and use those, but then when building for release packages we'd have to switch back to the normal ones.

tmakar replied:
> However, the bigger issue here is that this will not work, even if the PKG_CONFIG_PATH was set better in CMake. The issue is that pkg-config only finds an installed version of Elektra, it the pkgconfig files will never work properly for the version still being built. We'd have to generate special variants of the pkgconfig files and use those, but then when building for release packages we'd have to switch back to the normal ones.

As I am not an expert on this, I unfortunately didn't understand what you mean :smile:

tmakar replied:
> The immediate cause of the space issue is probably this. However, this line is not problematic and supports spaces properly ([golang/go#16455](https://github.com/golang/go/issues/16455)), as long as the pkg-config file is correct.

The error messages was a different one. This worked correctly but it then could not find other includes which were located under `....../src/includes`. In the specific case (on jenkins) it was `directory/src/includes` it could not found. The `build ` was omitted from the folder name and it therefore did not find the correct includes.

tmakar replied:
While trying to find the actual problem I also stumbled upon this https://github.com/golang/go/issues/16455.

tmakar commented:
@atmaxinger @kodebach @markus2330 when jenkins builds finish it can be merged!

kodebach commented:
I'm fine with removing the space in the Jenkins build dir for now, but if we merge it like this, we should create an issue to restore the space.

tmakar commented:
> I'm fine with removing the space in the Jenkins build dir for now, but if we merge it like this, we should create an issue to restore the space.

Will create an issue and link this PR.

tmakar commented:
jenkins build libelektra please

markus2330 commented:
there is a build error

kodebach commented on src/tools/elektrad/temp-elektra.pc.in:
```diff
@@ -1,10 +1,10 @@
 Version: @KDB_VERSION@
-Cflags: -I${srcinclude} -I${buildinclude}
-Libs: -L${libdir} -lelektra-kdb -lelektra-core
+Cflags: -I"\"${srcinclude}\"" -I"\"${buildinclude}\""
+Libs: -L"\"${libdir}\"" -lelektra-kdb -lelektra-core
```
```suggestion
Cflags: -I"${srcinclude}" -I"${buildinclude}"
Libs: -L"${libdir}" -lelektra-kdb -lelektra-core
```

For my tests this is enough. Like in a shell you only need to put quotes around the `${variable}`s. If I do do this with this test file locally:

```
prefix=test string abc
exec_prefix="${prefix}"/bin
libdir="${prefix}"/lib
includedir="${prefix}"/include/elektra
plugindir="${prefix}"/lib/elektra
tool_execdir="${prefix}"/lib/elektra/tool_exec
templatedir="${prefix}"/share/elektra/templates
build_type=shared

Name: 
Description: 
Version: 0.9.14
Libs: -L"${libdir}" -lelektra-kdb -lelektra-core
Cflags: -I"${prefix}"/include
```

And run `PKG_CONFIG_PATH=path/to/folder/of/file/above pkg-config --libs --cflags elektra` I get:

```
-Itest\ string\ abc/include -Ltest\ string\ abc/lib -lelektra-kdb -lelektra-core 
```

AFAICT this should be interpreted correctly by `cgo`. If that is not the case, this file is not the problem.

tmakar replied:
Yes, it worked locally as well. I had it this way, but jenkins was not happy again.

kodebach commented on src/bindings/go-elektra/temp.elektra.pc:
```diff
@@ -0,0 +1,10 @@
+Name: @PACKAGE_NAME@
+Description: @PACKAGE_DESCRIPTION@
+Version: @KDB_VERSION@
+Cflags: -I"\"${srcinclude}\"" -I"\"${buildinclude}\""
+Libs: -L"\"${libdir}\"" -lelektra
```
```suggestion
Cflags: -I"${srcinclude}" -I"${buildinclude}"
Libs: -L"${libdir}" -lelektra
```

See other comment

kodebach commented on src/bindings/go-elektra/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+	# build
+	add_custom_command (
+		OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/lib/libelektra.golib"
+		COMMAND ${CMAKE_COMMAND} ARGS -E env GO111MODULE=on ${GO_EXECUTABLE} build ./kdb
```
```suggestion
		COMMAND ${CMAKE_COMMAND} ARGS -E env PKG_CONFIG_PATH=${CMAKE_CURRENT_BINARY_DIR} GO111MODULE=on ${GO_EXECUTABLE} build ./kdb
```

To use the `elektra.pc` you generate above, you now need to add the `PKG_CONFIG_PATH` here.

kodebach commented:
@tmakar Please avoid even more force-pushes unless really necessary (e.g. because of rebase). It's better to have multiple redundant commits with the same message, if you're trying things. That way it's much easier to see on Github what you already tried. If it ends up being a lot of commits tried different things, you can always use an interactive rebase to squash the commit and then make a force-push at the end.

tmakar commented:
I have now found the following:

I explicitly changed the file contents of `elektra.pc` for `elektrad` to use `dummy libbbbbbb`:

```sh
prefix="@CMAKE_BINARY_DIR@"
libdir="dummy libbbbbbb/lib"
srcinclude="${prefix}/src/include"
buildinclude="@CMAKE_SOURCE_DIR@/src/include"

Name: elektra
Description: @PACKAGE_DESCRIPTION@
Version: @KDB_VERSION@
Cflags: -I"${srcinclude}" -I"${buildinclude}"
Libs: -L"${libdir}" -lelektra-kdb -lelektra-core
``` 

Then I ran it and got the following error:
```sh
github.com/ElektraInitiative/libelektra/go-elektra/kdb: invalid flag in pkg-config --libs: libbbbbbb/lib
make[2]: *** [src/tools/elektrad/CMakeFiles/elektrad.dir/build.make:78: src/tools/elektrad/elektrad/elektrad] Error 1
make[1]: *** [CMakeFiles/Makefile2:11687: src/tools/elektrad/CMakeFiles/elektrad.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

So, it has to be in this file and it might be fixed by enclosing `${prefix}` with quotes => `libdir="${prefix}"\lib`.

kodebach commented:
@tmakar I found the issue... The problem is the ancient `pkg-config` version (0.29.2) that Debian ships in the `pkg-config` package. It seems most other distros nowadays ship the alternative implementation `pkgconf` as their `pkg-config` version. That version properly supports spaces is what I was testing with (and you probably too).

So the solution is simply to change all our Debian Dockerfiles from installing `pkg-config` to `pkgconf`.

tmakar commented:
> @tmakar I found the issue... The problem is the ancient `pkg-config` version (0.29.2) that Debian ships in the `pkg-config` package. It seems most other distros nowadays ship the alternative implementation `pkgconf` as their `pkg-config` version. That version properly supports spaces is what I was testing with (and you probably too).
> 
> So the solution is simply to change all our Debian Dockerfiles from installing `pkg-config` to `pkgconf`.

:exploding_head: 

But how does it then also fail for fedora :thinking: ?

tmakar commented:
jenkins build libelektra please

tmakar commented:
jenkins build libelektra please

markus2330 commented:
jenkins build libelektra please

tmakar commented:
jenkins build libelektra please

tmakar commented:
So, `go` only knows about `pkg-config` and not about `pkgconf`, so I created a symlink for it in the dockerfiles, but unfortunately this does also not work :/

Here is the function go uses:
```go
// Calls pkg-config if needed and returns the cflags/ldflags needed to build the package.
func (b *Builder) getPkgConfigFlags(p *load.Package) (cflags, ldflags []string, err error) {
	if pkgs := p.CgoPkgConfig; len(pkgs) > 0 {
		var pcflags []string
		for len(pkgs) > 0 && strings.HasPrefix(pkgs[0], "--") {
			pcflags = append(pcflags, pkgs[0])
			pkgs = pkgs[1:]
		}
		for _, pkg := range pkgs {
			if !load.SafeArg(pkg) {
				return nil, nil, fmt.Errorf("invalid pkg-config package name: %s", pkg)
			}
		}
		var out []byte
		out, err = b.runOut(p.Dir, p.ImportPath, nil, b.PkgconfigCmd(), "--cflags", pcflags, "--", pkgs)
		if err != nil {
			b.showOutput(nil, p.Dir, b.PkgconfigCmd()+" --cflags "+strings.Join(pcflags, " ")+strings.Join(pkgs, " "), string(out))
			b.Print(err.Error() + "\n")
			return nil, nil, errPrintedOutput
		}
		if len(out) > 0 {
			cflags = splitPkgConfigOutput(out)
			if err := checkCompilerFlags("CFLAGS", "pkg-config --cflags", cflags); err != nil {
				return nil, nil, err
			}
		}
		out, err = b.runOut(p.Dir, p.ImportPath, nil, b.PkgconfigCmd(), "--libs", pcflags, "--", pkgs)
		if err != nil {
			b.showOutput(nil, p.Dir, b.PkgconfigCmd()+" --libs "+strings.Join(pcflags, " ")+strings.Join(pkgs, " "), string(out))
			b.Print(err.Error() + "\n")
			return nil, nil, errPrintedOutput
		}
		if len(out) > 0 {
			ldflags = strings.Fields(string(out))
			if err := checkLinkerFlags("LDFLAGS", "pkg-config --libs", ldflags); err != nil {
				return nil, nil, err
			}
		}
	}

	return
}
```

kodebach commented:
While in theory it should just work with `pkgconf` (no symlinks needed `pkgconf` adds `/use/bin/pkg-config`) it seems to need a lot more work than expected.

I just checked some more distros and:
- Fedora does have a pkgconf package, but AFAIK we are not installing any pkg-config there (who knows what it's using right now)
- Ubuntu (since it's a Debian derivative) probably also needs a different package
- OpenSUSE Leap only has the old `pkg-conf`

So we might even need some extra logic to set the build directory based on the distro.

Probably the best way forward here is:
1. Switch to a build directory without spaces on Jenkins
2. Create an issue for "build directory with spaces". There we can discuss what we do about it.

tmakar commented:
jenkins build libelektra please

markus2330 commented:
Let us see if this also builds on master.
